client/implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/HeaderBuilder.java

```diff
@@ -0,0 +1,117 @@
+package io.smallrye.graphql.client.typesafe.impl;
+
+import static io.smallrye.graphql.client.typesafe.impl.CollectionUtils.toMultivaluedMap;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Base64;
+
+import javax.ws.rs.core.MultivaluedMap;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.ConfigProvider;
+
+import io.smallrye.graphql.client.typesafe.api.AuthorizationHeader;
+import io.smallrye.graphql.client.typesafe.api.GraphQlClientApi;
+import io.smallrye.graphql.client.typesafe.api.GraphQlClientException;
+import io.smallrye.graphql.client.typesafe.api.Header;
+import io.smallrye.graphql.client.typesafe.impl.reflection.MethodInfo;
+import io.smallrye.graphql.client.typesafe.impl.reflection.MethodResolver;
+import io.smallrye.graphql.client.typesafe.impl.reflection.TypeInfo;
+
+public class HeaderBuilder {
+    private final MethodInfo method;
+
+    public HeaderBuilder(MethodInfo method) {
+        this.method = method;
+    }
+
+    public MultivaluedMap<String, Object> build() {
+        MultivaluedMap<String, Object> headers = method.getResolvedAnnotations(Header.class)
+                .map(header -> new SimpleEntry<>(header.name(), resolveValue(header)))
+                .collect(toMultivaluedMap());
+        method.parameters()
+                .filter(parameter -> parameter.isAnnotated(Header.class))
+                .forEach(parameter -> {
+                    Header header = parameter.getAnnotations(Header.class)[0];
+                    headers.add(header.name(), parameter.getValue());
+                });
+        method.getResolvedAnnotations(AuthorizationHeader.class)
+                .findFirst()
+                .map(header -> resolveAuthHeader(method.getDeclaringType(), header))
+                .ifPresent(auth -> headers.add("Authorization", auth));
+        return headers;
+    }
+
+    private Object resolveValue(Header header) {
+        if (!header.method().isEmpty()) {
+            if (!header.constant().isEmpty())
+                throw new GraphQlClientException("Header with 'method' AND 'constant' not allowed: " + header);
+            return resolveMethodValue(header.method());
+        }
+        if (header.constant().isEmpty())
+            throw new GraphQlClientException("Header must have either 'method' XOR 'constant': " + header);
+        return header.constant();
+    }
+
+    private Object resolveMethodValue(String methodName) {
+        TypeInfo declaringType = method.getDeclaringType();
+        MethodInfo method = new MethodResolver(declaringType, methodName).resolve();
+        if (!method.isStatic())
```

Comment:
MP Rest Client does something similar with annotation-based header declaration (`@ClientHeaderParam`).  It allows users to call static methods on any accessible class or default method on the interface itself.  That might be something to consider here for consistency.

Reply:
I had taken a look at the implementation, and it's quite complex. I still want to take a closer look into the whole `MethodHandle` API, so I wanted to start more restrictive and open it up later.

Reply:
I looked into MethodHandles but I'm really not sure if it's worth the hassle to call `default` methods. Users can declare their methods to be `static` methods just as well (hey, it's even one char less ;-) I'm looking for use-cases for default methods, but even when testing code accessing the API interface by mocking it: there's not need to override the method, as it won't be called.

Maybe there is a use-case I don't see?

Reply:
Another update: it's allowed to access private or protected methods in enclosing classes, even without qualifying it with the class name; this is a more common use-case. MP Rest Client doesn't support that, though. I'm not sure how to align them. What do you think?

Reply:
I think that is a good approach.  We can always open it up later - but it's a lot harder to restrict it later if we find out that it doesn't add much value.